### ISSUE TITLE
User Settings; Disable certain software

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#37]
Modular system to have per-user settings to disable certain executables / group of executables.
Something simplistic to start with. 

Due to rampant abuse on Storage Boxes, we need to disable all torrent clients except aria2c.
Same with CPU Miners, and MTAs.

So 2 sets of configs, global all users, and per user settings.
Global by default has rules to disable running MTA, CPU Miners, and then account specific to disable torrent clients.

If abuse continues we'll tighten rules.

**Comment by MagnaCapax** [comment]
## Investigation Report

**Issue:** #37 — User Settings; Disable certain software
**Investigated:** 2026-05-31
**Status:** PARTIAL

### Summary
This issue bundles three distinct needs under one "modular executable-disable engine" framing. On analysis, two of the three (CPU miners, MTAs) are better addressed by mechanisms PMSS already has or can add trivially (cgroup CPU caps; an outbound port‑25 egress rule), and the third (torrent clients) is a declarative per‑user *service toggle* that already largely exists. A generic name‑matching kill engine is the wrong abstraction. **Recommendation: SPLIT_FIRST.**

### What The Issue Asks For
A modular per‑user + global config to disable executables/groups, "something simplistic to start with," driven by Storage Box abuse:
- Per‑account: disable all torrent clients **except aria2c**.
- Global default: disable CPU miners and MTAs.
- Two config tiers: global (all users) and per‑user.

### Implementation Status
The disable/toggle domain is **partially built**:

- **Per‑user service‑toggle pattern** routes through `pmssUserConfigNormaliseToggleValue()` in `scripts/lib/user/userConfigStore.php` (normalise ~`:267-268`). Helpers `pmssUserDockerEnabled` (~`:361`, RAM‑floor gate) and `pmssUserLighttpdEnabled` (~`:386`, pure toggle) exist.
- **Watchdog enable‑marker (dotfile) pattern** at `scripts/lib/user/watchdog.php:157-169` — `pmssUserWatchdogRunService($heading, $enableMarker, $processNames, …)` skips startup and stops running processes when `/home/<user>/.<enableMarker>` is absent. Deluge uses `delugeEnable` (`scripts/cron/checkDelugeInstances.php:14`); qBittorrent uses `qbittorrentEnable` (`scripts/cron/checkQbittorrentInstances.php:10`).
- **So torrent‑client disable is already mostly possible**: deluge/qBittorrent via dotfile; rtorrent via the pending toggle in **#470** (confirmed: `rtorrentEnabled` has **zero** occurrences repo‑wide as of 2026‑05‑31). `aria2c` has no watchdog (packaged binary only), so "except aria2c" needs nothing.
- **CPU‑cap mechanism** exists (`cpuQuotaPercent` via the cgroup manager / per‑user config) **but is not applied by default** — shipped `etc/seedbox/config/cgroup.policy.php` sets only `cpuWeight => 100` (a *relative* share under contention), not a hard quota.
- **Process‑name kill precedent**: `scripts/cron/processWatchdog.sh` kills `ffmpeg`/`HandBrakeCLI` by `comm` with a two‑strike policy — hardcoded, not config‑driven, not per‑user.

### What's Missing
- A landed rtorrent toggle (#470).
- A hard CPU cap applied by default on the Storage Box product (mechanism present; default policy doesn't set it).
- An outbound TCP/25 egress block — **none exists**: the firewall `OUTPUT` chain defaults to `ACCEPT`; `scripts/util/setupNetwork.php` defines INPUT/FORWARD restrictions and overlay‑egress ACCEPTs only, with no `--dport 25` rule anywhere in `scripts/`.
- A "global all users" *service‑disable* config tier (only `cgroup.policy.php` is fleet‑wide, and it governs resources, not service enablement).

### Why not a generic "disable executables" engine
- **Name‑based detection is evadable** — renaming a miner binary or running a custom SMTP client under any name defeats a `comm`‑name match, so it gives false assurance against the very adversaries it targets, and *fails open* when evaded.
- It would require **elevated privilege** to police arbitrary user processes and adds a new failure mode (false‑positive kills of legitimate work).
- The OS layer that would properly own executable control (SELinux/AppArmor) was **explicitly deferred** in **ADR‑0011** for stability reasons; a bespoke name‑killer is a weaker reinvention of it.

By contrast: a cgroup `cpuQuotaPercent` cap makes mining unprofitable *without detecting it* ("a miner is just a user using their capped share"); an outbound‑25 block stops spam connections regardless of which binary opens them. Both *fail closed*.

### Cross‑Issue Analysis
- **#470 (OPEN) — building block / DEPENDS.** Implements the rtorrent per‑user toggle = the rtorrent slice of this issue's torrent‑toggle track. Land it; do not re‑implement here.
- **#270 (CLOSED) — precedent.** Per‑product‑type docker default; precedent for product‑scoped defaults (relevant to Storage‑Box‑scoped policy).
- **#157 (via ADR‑0011) — related.** The deferred MAC request — the "proper" executable‑control layer this issue informally reaches for.
- **#437 (OPEN) — related.** cgroup throttle + OOM‑kill logging; adjacent to the miner→CPU‑cap path (observability for caps).
- **#121 (OPEN) — related.** Pending `cgroup.policy` extensions; a natural home for a default Storage‑Box `cpuQuotaPercent`.
- No single open issue supersedes #37 — it is the umbrella; #470 is one slice.

### ADR Adherence
- **Adheres to ADR‑0011 (defer default MAC enforcement):** the recommendation declines to build a bespoke executable‑control engine, consistent with deferring the layer that would properly own it; abuse is routed to lighter targeted controls.
- **Adheres to ADR‑0004 (shell guardrails) / ADR‑0005 (trust boundaries):** the toggle + config + firewall approach involves no untrusted‑data shell construction and keeps enforcement in operator‑side code.
- **Adheres to ADR‑0016 (customer/operator tree separation):** all touched surfaces (watchdogs, config store, firewall, cgroup policy) are operator‑side `scripts/`.
- A **new ADR is proposed** (below) to record "prefer targeted controls over a generic name‑based process‑disable engine"; it extends ADR‑0011 rather than conflicting with it.

### Dependencies & Blockers
- The rtorrent slice depends on #470 landing.
- The port‑25 egress block needs an operator decision on scope (fleet‑wide vs Storage‑Box product) — full‑seedbox/dedicated customers may legitimately need outbound 25, so an allowlist hook is advisable.
- A global service‑disable tier (if ever pursued) is ADR‑level work — but the decomposition argues it is unnecessary.

### Reuse, Refactor & Test Strategy
- **Reuse**: the toggle normaliser + helpers in `userConfigStore.php`; the dotfile watchdog framework `pmssUserWatchdogRunService`; `lighttpdEnabled` as the exact template (per #470); `userSetting.php set USER <key> <val>` as the operator API (no new CLI needed); `cgroup.policy.php` for the CPU cap; `setupNetwork.php` for the egress rule.
- **Refactor**: converge the two enable/disable patterns (dotfile markers vs JSON keys) onto one — the JSON store is the crystallizing choice. This is the one genuine architectural cleanup the issue surfaces.
- **Tests**: unit (toggle default‑true + explicit‑false in `userConfigStoreTest.php`); characterization (assert each torrent watchdog consults its toggle, mirroring `cronInlineCharacterizationTest.php`); regression (a low‑RAM seedbox user who *uses* rtorrent must not be auto‑disabled — i.e. a pure toggle, no RAM‑floor auto‑disable, per #470); boundary (egress rule must not break loopback/overlay mail); failure‑injection (renamed‑miner case, to document why detection is *not* the chosen control).
- **Minimum code for correctness**: extend the existing toggle pattern to the remaining torrent watchdogs; one `cgroup.policy` default; one firewall line. No new engine, no global‑tier abstraction, no name‑matching rules.

### Recommendation
**SPLIT_FIRST** — decompose into per‑track verdicts rather than building #37 as one unit:
1. **Torrent per‑user toggles → IMPLEMENT.** Mirror the `lighttpdEnabled`/#470 pattern across rtorrent (#470)/deluge/qBittorrent; expose via existing `userSetting.php`. Low risk, mostly patterned, customer‑driven.
2. **CPU miners → apply the existing `cpuQuotaPercent` cap** on Storage Box defaults (config change). No detection engine.
3. **MTAs → outbound TCP/25 egress block** (structural firewall rule) — see new‑issue suggestion.
4. **Drop the generic engine + global service‑disable tier → do not implement that sub‑part:** premature, relies on evadable name‑matching, and reinvents the layer deferred by ADR‑0011.

**Suggested follow‑ups:**
- **New issue:** *"Block outbound TCP/25 (SMTP) egress to curb MTA abuse"* — add a `--dport 25 -j REJECT` rule in the OUTPUT path (`scripts/util/setupNetwork.php`); operator to decide fleet‑wide vs Storage‑Box‑scoped, with an allowlist for legitimate senders.
- **New ADR:** *"Prefer targeted controls over a generic name‑based process‑disable engine"* — record that PMSS meets the "disable software" need via declarative service toggles, cgroup resource caps, and firewall egress filtering, rather than an evadable process‑name killer that overlaps the deferred MAC layer.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Independent verification of the prior analysis (against current `main`)

I re-checked every load-bearing source claim in the investigation report above against current `main` (HEAD `df0c5b9e`, 2026-05-31). **All hold — no drift:**

- **Per-user service-toggle pattern present** — `pmssUserConfigNormaliseToggleValue()` (`scripts/lib/user/userConfigStore.php:53`); `dockerEnabled`/`lighttpdEnabled` normalised at `:267-268`; helpers `pmssUserDockerEnabled` (`:361`) / `pmssUserLighttpdEnabled` (`:386`).
- **Watchdog dotfile markers present** — `delugeEnable` (`scripts/cron/checkDelugeInstances.php:14`), `qbittorrentEnable` (`scripts/cron/checkQbittorrentInstances.php:10`), via `pmssUserWatchdogRunService` (`scripts/lib/user/watchdog.php:157`).
- **`rtorrentEnabled`: zero occurrences repo-wide** (the pending toggle in #470). `aria2c`: no watchdog (packaged binary only) → "except aria2c" needs nothing.
- **`cgroup.policy.php` default sets only `cpuWeight => 100`; no hard `cpuQuotaPercent`** — the CPU-cap mechanism exists but is not applied by default (confirms the report's self-correction).
- **Outbound TCP/25 egress block absent** — zero `dport 25` rules repo-wide.
- **`processWatchdog.sh:67` hardcodes `comm` ∈ {`ffmpeg`, `HandBrakeCLI`}** with a two-strike kill — not config-driven, not per-user.
- **ADR-0011 = "Defer default SELinux/AppArmor enforcement"** (refs #157) — read correctly.

**The SPLIT_FIRST recommendation stands and is now independently confirmed.** Concrete next product steps:

1. **Torrent toggles** — rtorrent slice is tracked by **#470**; deluge/qBittorrent already disable via their dotfile markers; expose via `userSetting.php`.
2. **CPU miners** — apply the existing `cpuQuotaPercent` cap on Storage Box defaults (config change). Natural home: **#121 / #437**.
3. **MTAs** — a new issue for an outbound TCP/25 egress rule in `scripts/util/setupNetwork.php` (operator to decide fleet-wide vs Storage-Box scope, with an allowlist for legitimate senders).

No generic name-based "disable engine": it is evadable by rename (fails open) and overlaps the SELinux/AppArmor layer that ADR-0011 deliberately defers. The three targeted controls (service toggle, cgroup cap, firewall egress) each fail closed and sit at the correct layer.

Since the analysis is now converged and verified, this issue is best resolved by routing the three tracks above (one is #470; one folds into #121/#437; one new issue) and then closing #37 as decomposed.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Investigation Report — 3rd review (corrections + disposition)

**Issue:** #37 — User Settings; Disable certain software
**Investigated:** 2026-06-02
**Status:** PARTIAL

> This is the **third** investigation of this issue. The two prior reports above (2026-05-31) converged on **SPLIT_FIRST** and verified each other. To avoid re-deriving converged analysis, this review does only three things: **(1) corrects one load-bearing routing error in the prior reports, (2) adds a missed MTA precedent, and (3) gives the exact disposition to close this issue.** The technical verdict is unchanged — the bottleneck now is disposition, not analysis.

### Summary
#37 bundles three abuse-control needs (torrent clients, CPU miners, MTAs) under a "generic per-user/global executable-disable engine." That engine should **not** be built (name-based detection is evadable, fails open, needs privilege, and reinvents the MAC layer deferred by ADR-0011). The real needs decompose into three targeted, fail-closed controls. Re-verified against current `main` (HEAD `6d0dd47b`, 2026-06-01): **no drift.** This review corrects the prior "CPU miners fold into #121/#437" claim (they don't) and notes the system-MTA slice is already done (#277/#275).

### What The Issue Asks For
A modular per-user + global config to disable executables/groups ("something simplistic to start with"), driven by Storage Box abuse: per-account disable all torrent clients **except aria2c**; global default disable CPU miners and MTAs; two config tiers (global + per-user).

### Implementation Status (verified at HEAD `6d0dd47b`)
- **Per-user service-toggle pattern** — present: `pmssUserConfigNormaliseToggleValue()` (`scripts/lib/user/userConfigStore.php`), helpers `pmssUserDockerEnabled` / `pmssUserLighttpdEnabled`.
- **Torrent disable already partly works.** deluge & qBittorrent are **opt-in / off by default** via enable-marker dotfiles — `scripts/lib/user/watchdog.php:169` skips a user whose `.<marker>` file is absent (`delugeEnable` / `qbittorrentEnable`). Provisioning never creates these markers (only test fixtures do), so absence = disabled. `aria2c` has no watchdog → "except aria2c" needs nothing.
- **rtorrent** auto-regenerates and has no toggle — `rtorrentEnabled` still **zero occurrences repo-wide**. This is the slice tracked by **#470**.
- **CPU-cap mechanism exists but is not a default** — shipped `etc/seedbox/config/cgroup.policy.php` sets `cpuWeight => 100` (a *relative* share), no hard `cpuQuotaPercent`. A miner on an idle host hits no limit.
- **System MTA already eliminated** — `exim4` is purged (`#277` CLOSED) + spool purge & `MAILTO=""` (`#275` CLOSED, complete-verify). The "disable MTA" need is met for the *system* MTA.
- **Name-based kill precedent** — `scripts/cron/processWatchdog.sh:67` hardcodes a two-strike kill for `ffmpeg`/`HandBrakeCLI` by `comm`; not config-driven, not per-user.

### What's Missing
- A landed rtorrent toggle (**#470**).
- A **default CPU quota** on the Storage Box product (mechanism present; default policy doesn't set it). **No issue currently tracks this** — see correction below.
- An **outbound TCP/25 egress block** for customer-run SMTP — zero `dport 25` rules repo-wide. (System MTA is gone; the residual is *customer processes* sending spam.)
- A "global all-users service-disable" tier — the decomposition argues this is unnecessary (resource policy already fleet-wide; service enablement is per-user).

### Corrections to the prior reports
1. **The CPU-miner track is UNROUTED.** Prior reports said it "folds into #121/#437." It does not: **#121** is IO/burst/net/NOFILE cgroup TODOs; **#437** is observability-only (memory.events / OOM / watchdog logging). Reading both bodies, neither mentions a CPU cap/quota, a miner, or a Storage-Box default. If #37 were closed trusting the prior reports, this track would be silently dropped. It needs its **own** issue.
2. **The MTA track is half-done, not wholly open.** Prior reports framed "MTA → new port-25 issue" as fully unaddressed. In fact **#277 / #275 (both CLOSED)** already removed the system MTA (exim4) and suppressed cron mail; #275's own security note flags the open-relay/abuse vector. Only the **customer-process SMTP egress** slice remains.

### Why not a generic "disable executables" engine
Name/`comm`-based detection is defeated by a rename (fails **open** against the very adversaries it targets), needs elevated privilege to police arbitrary user processes, and adds a false-positive-kill failure mode. The OS layer that would properly own this (SELinux/AppArmor) was **deliberately deferred in ADR-0011**. By contrast a `cpuQuotaPercent` cap makes mining unprofitable *without detecting it*, and an outbound-25 block stops spam regardless of binary name — both fail **closed**, at the correct layer.

### Cross-Issue Analysis
- **#470 (OPEN) — building block / DEPENDS.** The rtorrent slice of the torrent track. Land it; do not re-implement here.
- **#277 / #275 (CLOSED) — the MTA system-default slice, already done.** exim4 purged + cron mail suppressed.
- **#121 (OPEN) / #437 (OPEN) — NOT the miner home** (corrected). #121 = other cgroup extensions; #437 = observability. The miner CPU-cap track has no tracking issue.
- **#157 (via ADR-0011) — related.** The deferred MAC layer this issue informally reaches for.
- **#270 (CLOSED) — precedent** for per-product-type defaults (Storage-Box-scoped policy).
- **#378 (OPEN) — explains why this keeps getting re-selected** (its candidate-pool filter excludes only `complete-verify`/`wontfix`/`needs-investigation`, none of which #37 carries) but its proposed fix would not exclude #37, whose verdict is SPLIT_FIRST.
- **No single open issue subsumes #37** — it is the umbrella; #470 is one of three tracks.

### Dependencies & Blockers
rtorrent slice depends on #470 landing. The CPU cap needs an operator-chosen quota % (mechanism ready). The port-25 egress block needs an operator scope decision (fleet-wide vs Storage-Box product; allowlist for legitimate senders).

### Reuse, Refactor & Test Strategy
- **Reuse:** the toggle normaliser + `lighttpdEnabled` helper as the template for #470; the dotfile watchdog framework `pmssUserWatchdogRunService`; `userSetting.php set USER <key> <val>` as the operator API (no new CLI); `cgroup.policy.php` for the cap; `setupNetwork.php` for the egress rule.
- **Refactor:** converge the two enable/disable patterns (dotfile markers vs JSON keys) onto the JSON store — the one genuine architectural cleanup #37 surfaces.
- **Tests:** unit (toggle default-true + explicit-false); characterization (each torrent watchdog consults its toggle); regression (a low-RAM user who *uses* rtorrent must not be auto-disabled — pure toggle, no RAM-floor, per #470); boundary (egress rule must not break loopback/overlay mail); failure-injection (renamed-miner case — documents why detection is not the chosen control).
- **Minimum code for correctness:** the rtorrent toggle (#470), one `cgroup.policy` default, one firewall line. No engine, no global-tier abstraction.

### Architectural Minimization
- **Question the requirement:** "build a *modular executable-disable engine*" presumes process-level enforcement is the tool. It was written in 2024 before the per-user toggle pattern and cgroup policy existed, and conflates three needs at the wrong layer. The real requirement is "stop torrent abuse on Storage Boxes, stop miners, stop spam," not "build an engine."
- **DELETE (wins):** delete the generic-engine idea; its function is covered by existing service toggles + the existing `cpuQuotaPercent` cap + a firewall rule.
- **MERGE:** rtorrent slice merges into #470; the MTA system slice already merged into #277/#275.
- **REFACTOR:** converge dotfile-marker and JSON-key toggle patterns into the JSON store.
- **NULL:** if nothing is done, rtorrent abuse is unconstrained until #470 and miners/customer-SMTP remain weakly constrained — so the *tracks* are worth doing; the *engine* is not.

### ADR Adherence
- **ADHERES: ADR-0011 (defer default SELinux/AppArmor)** — declining a bespoke executable-control engine is consistent with deferring the layer that would own it.
- **ADHERES: ADR-0004 / ADR-0005** — toggle + config + firewall involves no untrusted-data shell construction.
- **ADHERES: ADR-0016 (customer/operator tree separation)** — all touched surfaces are operator-side `scripts/`.
- A new ADR is *proposed* below, extending ADR-0011; no conflict.

### Recommendation
**SPLIT_FIRST** (affirmed — third convergence; this should be the final review). The literal "modular executable-disable engine + global service-disable tier" is **SHOULD-NOT-IMPLEMENT**. Resolve #37 by routing three tracks, then closing it as decomposed:
1. **Torrent toggles → IMPLEMENT.** rtorrent via **#470**; deluge/qBittorrent already off by default; `aria2c` needs nothing.
2. **CPU miners → file a new issue** to apply a default `cpuQuotaPercent` on the Storage Box product (operator picks the %). **Do not** rely on #121/#437 — neither covers it.
3. **MTAs → file a new issue** for an outbound TCP/25 egress block (system MTA already handled by #277/#275; operator scopes fleet-wide vs Storage-Box, with a sender allowlist).
4. **Generic engine + global tier → do not implement.**

**Disposition (to stop the re-selection loop):** once the two new issues exist and #470 is tracked, **close #37 as decomposed** (or apply a disposition label now). This issue's analysis is complete; further investigation passes add nothing.

### Suggested follow-up issues
- **New issue:** *"Apply a default cgroup CPU quota (`cpuQuotaPercent`) on the Storage Box product to curb CPU-miner abuse"* — set a hard cap in the product's cgroup policy (mechanism exists); operator chooses the %.
- **New issue:** *"Block outbound TCP/25 (SMTP) egress to curb customer-process MTA/spam abuse"* — add a reject rule in the OUTPUT path (`scripts/util/setupNetwork.php`); operator decides fleet-wide vs Storage-Box, with a legitimate-sender allowlist. (System MTA already removed by #277/#275.)
- **New ADR:** *"Prefer targeted controls over a generic name-based process-disable engine"* — record that #37's "disable software" need is met by declarative service toggles, cgroup resource caps, and firewall egress filtering (each failing closed at the correct layer), rather than an evadable, privileged process-name killer that overlaps the layer ADR-0011 defers.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review

**Issue:** #37 — User Settings; Disable certain software
**Reviewed:** 2026-06-13
**Verdict:** SHOULD-NOT-IMPLEMENT (for the literal ask: a generic name-based executable-disable engine + a global service-disable tier)

> This is the **first adversarial review** to reach this issue. Three investigation reports above (2026-05-31 ×2, 2026-06-02) converged on decomposition. To re-derive that a fourth time would itself be the anti-pattern. So this review does only what is still missing: **(1) one new adversarial finding the prior passes under-stated, (2) a crisp build/skip verdict, and (3) the disposition action that actually ends the re-selection loop — the two unrouted tracks are now filed (#660, #661), so the umbrella can be closed without dropping work.**

### Adversarial Findings

- **[REGEX_FOR_SEMANTICS] The engine classifies process *intent* by binary *name*.** "Disable CPU miners / MTAs / torrent clients" by matching `comm` is judging semantics with string equality. `scripts/cron/processWatchdog.sh:67` is the existing precedent (`comm` ∈ {`ffmpeg`,`HandBrakeCLI`}, two-strike kill) — and it is exactly the shape that fails the moment a binary is renamed.
- **[FAIL_OPEN_SAFETY_BEHAVIOR] A name-matcher fails *open* against the very adversary it targets.** A miner renamed to `httpd`, or a custom SMTP client under any name, evades the match and runs free — while giving operators false assurance that "miners are disabled." The control is weakest precisely where abuse is most deliberate.
- **[NEW — not in prior passes] The per-user service toggles are a *convenience* mechanism, not an *enforcement* boundary for active abusers.** The enable marker is read at `scripts/lib/user/watchdog.php:179` as `is_file('/home/<user>/.<marker>')` — inside the **user-writable home directory**, and the suspended-user path is handled separately first (`pmssUserWatchdogHandleSuspended`). So for an *active* (non-suspended) Storage Box user — which is #37's actual driver — "disable" via a home-dir dotfile is not enforceable: the user re-creates the marker, or simply launches the base-installed binary (`scripts/lib/update/apps/rtorrent.php`, `apps/deluge.php`) directly. The toggles control whether PMSS *auto-starts* a service (real value: RAM saving, per #470), not whether a determined user *can run* it. The three prior reports framed "deluge/qBittorrent disable already works today" without this distinction; it matters because it means **none of the per-user toggles actually solve the stated abuse problem** — only structural controls do.
- **Worst case if implemented as described:** operator ships a `comm`-based per-user/global kill engine → it needs elevated privilege to police arbitrary user processes → it false-positive-kills a legitimate long-running job (the `ffmpeg` precedent already kills by name) → customer data/work is lost mid-flight → complaint arrives hours later → manual recovery, trust damage. The engine adds a *new* catastrophic failure mode while still failing open on real miners.
- **Inversion (guarantee maximum damage):** give an unprivileged, evadable name-matcher root authority to SIGKILL user processes fleet-wide on a shared list. The proposed engine is one careless rule away from that.
- **Regret in 6 months:** a fleet-wide name list nobody maintains, drifting out of date, occasionally killing legitimate work, providing security theatre against miners that were renamed on day one.
- **Removal inversion:** if this engine already existed, would we fight to keep it? No — we would replace it with cgroup caps and egress rules (controls that fail closed) and delete the name list. That is the signal it should not be added.
- **The dumbest requirement:** "build a *modular executable-disable engine*." It presumes process-identity enforcement is the tool. The real requirement is "stop torrent/miner/spam abuse on Storage Boxes" — which is a *resource and egress* problem, solved at the cgroup and firewall layers, not a *process-name* problem.
- **Deletion opportunity:** delete the engine concept entirely. Its function is already covered by service toggles (convenience) + cgroup resource caps (`etc/seedbox/config/cgroup.policy.php`, `cpuQuotaPercent`) + outbound traffic limits (`scripts/cron/trafficLimits.php`) + a firewall egress rule. PMSS's abuse-control architecture is *already* structural; a name-killer contradicts that direction.

### Caller Analysis

Not applicable in the removal sense — #37 *adds* a new engine; it removes no existing code, so there are no callers to enumerate for a removal. The adjacent surface it would touch (the enable/disable toggle pattern) has known consumers that any work must preserve: `scripts/cron/checkDelugeInstances.php:15` (`delugeEnable`), `scripts/cron/checkQbittorrentInstances.php:10` (`qbittorrentEnable`), both via `pmssUserWatchdogRunService` (`scripts/lib/user/watchdog.php:167`), and the JSON store helpers (`pmssUserDockerEnabled`/`pmssUserLighttpdEnabled` in `scripts/lib/user/userConfigStore.php`). The pending rtorrent slice (`rtorrentEnabled`, **zero occurrences repo-wide** — confirmed) is #470.

### Investigation Results

Re-verified every load-bearing claim of the prior reports against current `main` (HEAD `87c52132`, 2026-06-12). **All hold, no drift:**

- Toggle pattern present — `pmssUserConfigNormaliseToggleValue()` normalises `dockerEnabled`/`lighttpdEnabled` (`scripts/lib/user/userConfigStore.php:223-224`).
- Watchdog dotfile markers present and user-home-scoped (`scripts/lib/user/watchdog.php:179`).
- `rtorrentEnabled`: zero occurrences → the #470 slice. `aria2c`: no watchdog → "except aria2c" needs nothing.
- `etc/seedbox/config/cgroup.policy.php` sets `cpuWeight => 100` (relative share) and **no hard `cpuQuotaPercent`** — the CPU cap mechanism exists (`scripts/lib/user/userConfigCli.php`, `--cpu-quota-percent`) but is not a Storage-Box default.
- Outbound TCP/25 egress block absent — `scripts/util/setupNetwork.php` OUTPUT chain has overlay ACCEPTs only, zero `dport 25` rules repo-wide.
- System MTA already removed — `exim4` purged (`scripts/lib/update/services/systemd.php:219`; `scripts/lib/support/mail.php` documents it) per #277/#275. Only the customer-process SMTP slice remains.

ADR adherence: declining a bespoke executable-control engine is consistent with ADR-0011 (default SELinux/AppArmor enforcement deferred — that is the layer this issue informally reaches for). All touched surfaces are operator-side `scripts/` (ADR-0016).

### Recommendations

- **Do not build the generic engine or the global service-disable tier.** Route the three real needs instead — each fails *closed*, at the correct layer:
  1. **Torrent track → #470** (rtorrent toggle; deluge/qBittorrent already off-by-default as a RAM-saving convenience). Note this is convenience, not abuse-enforcement (see finding above).
  2. **CPU-miner track → #660** (newly filed) — apply a default `cpuQuotaPercent` on the Storage Box product. Makes mining unprofitable without detecting it. #121/#437 do **not** cover this; it was genuinely unrouted until now.
  3. **MTA/spam track → #661** (newly filed) — outbound TCP/25 egress block. System MTA already handled by #277/#275; this is the residual customer-process slice.
- **Disposition:** with all three tracks now routed (#470/#660/#661), the operator can **close #37 as decomposed**. This issue carries no terminal label, which is why the candidate filter keeps re-selecting it; the `should-not-implement` label is applied to the *literal engine ask* so the Build agent skips it. The label does **not** reject the three routed tracks — those proceed on their own issues.

### Worth-Implementing Analysis

- **Cost:** the engine is high — new privileged process-policing daemon, a maintained name list, per-user + global config plumbing, and a new false-positive-kill failure mode forever.
- **Benefit:** low and illusory — evadable by rename, so it does not stop a deliberate abuser; it only catches the lazy case the existing `processWatchdog.sh` shape already half-covers.
- **Verdict rationale:** SHOULD-NOT-IMPLEMENT because a deletion alternative exists and is strictly better (structural caps + egress filtering), the mechanism fails open against its target, and it reinvents the deferred MAC layer. The legitimate sub-needs are real and now individually tracked.
- **Alternative:** the three routed tracks above.

### Merit Check — Tier A

1. **Concrete present demand.** Real, from the issue body itself: "Due to rampant abuse on Storage Boxes, we need to disable all torrent clients except aria2c." The *demand* (stop abuse) is genuine; the *proposed mechanism* (a name-based engine) is what fails. Demand for the engine specifically: none beyond this framing.
2. **Function.** As literally described, it would scan user processes by binary name against per-user + global disable lists and kill/prevent matches — a privileged `comm`-matching killer generalising `processWatchdog.sh`.
3. **Success metric.** None definable that is attributable to the engine: because a renamed miner evades it, a drop in miner load cannot be credited to the engine vs. to cgroup caps or manual escalation. Undefinable success metric is itself a SHOULD-NOT-IMPLEMENT signal.
4. **Catastrophic if implemented.** Privileged name-matcher gains SIGKILL authority over user processes → a rule matches a legitimate long-running job (the existing `ffmpeg` killer is precedent) → user's in-flight work dies silently → complaint surfaces hours later → per-user manual recovery + trust loss; meanwhile real miners renamed around it run unimpeded (false assurance compounds the damage).
5. **Catastrophic if NOT implemented.** Minimal and bounded: abuse continues *only* in the gap left by the structural controls — and those gaps are exactly #660 (CPU cap) and #661 (egress), which close them at the right layer. The asymmetry (large downside if built, small bounded downside if skipped) is the verdict.
6. **Deletion alternative.** Yes — delete the engine concept; its function is met by existing service toggles + `cpuQuotaPercent` caps + outbound traffic limits + one firewall egress rule. This is the preferred path.
7. **Cost.** (a) Build: high — new privileged daemon + config tiers + name catalog, well past a one-commit change. (b) Maintenance/year: the binary-name catalog needs perpetual curation as miners/clients rename — a standing tuning burden with no natural owner. (c) Blast radius if it breaks: any user running a name-matched process on shared hosts; false-positive kills hit legitimate work. (d) Undo: removing a shipped fleet-wide kill engine is a behavioural rollback across hosts, not a clean `git revert`.
8. **Dependency blast.** Nothing depends on the engine (it does not exist). The adjacent toggle pattern it would touch is consumed by the deluge/qBittorrent watchdogs and the JSON config helpers named in Caller Analysis; the rtorrent slice (#470) is the only pending consumer. Any toggle work must preserve those.
9. **Requirements challenge.** The dumbest requirement is "build a *modular executable-disable engine*," because it encodes a process-identity solution to a resource-and-egress problem. Strip it and the requirement becomes "stop miner CPU theft and spam egress on Storage Boxes" — answered structurally by #660 and #661.

**Tier B triggered — Goodhart (adds a policy/threshold):** optimising "number of disabled binaries" incentivises a growing brittle name list that looks protective while abusers rename around it — the metric improves as real protection does not. **Tier B triggered — Gall/second-system (proposes a generic engine over working primitives):** the incremental path being skipped is "extend the proven per-service toggle + cgroup cap + firewall primitives"; the engine smuggles in fleet-wide process policing far beyond the three concrete needs.

### Anti-Pattern Hunt

- **REGEX_FOR_SEMANTICS: FAIL** — the engine judges process intent by `comm` name-matching (`processWatchdog.sh:67` is the precedent). This sets the verdict floor at SHOULD-NOT-IMPLEMENT.
- **PREMATURE_OPTIMIZATION: FAIL (as a feature, not an optimisation)** — Exists: no. Tested: no. Useful: no (evadable, fails open). Worth-it-now: no (needs privilege, adds false-positive-kill mode). → SHOULD-NOT-IMPLEMENT for that part.
- **OPERATOR_HANDOFF_ANTIPATTERN: PASS** — the review does not punt unresolved work to an alert/handoff queue; it *closes capability* by filing the two structural tracks (#660/#661) and naming a concrete close-path. The only items left to the operator are genuine scope/value decisions (quota %, egress scope), correctly labelled `decision-needed`.
- **FALSE_COMPLETION_CLAIM: PASS** — no "fixed/done" claimed; the engine is explicitly *not* built, and the routed tracks are stated as pending on their own issues.
- **FABRICATED_AUTHORITY_OR_PATH: PASS** — every cited path/line was read this session against HEAD `87c52132`; ADR-0011/0016 read directly.
- **FAIL_OPEN_SAFETY_BEHAVIOR: FAIL (for the proposed engine)** — name-matching fails open on rename; this is a reason to reject it. The three routed controls fail closed (toggle state, cgroup cap, egress block).
- **DRY_RUN_THEATER: PASS** — N/A; read-only review, no dry-run substituted for evidence.
- **INVESTIGATION_LOOP_THEATER: PASS (explicitly avoided)** — this review does not re-derive the converged analysis; it adds one new finding and executes the disposition (files the unrouted tracks, applies the terminal label) that converts four passes of recommendation into action. A further substantive re-investigation WOULD fail this check.
- **LACK_OF_INVESTIGATION: PASS** — two counter-signals named and queried: (1) "is the dotfile toggle enforceable for active users?" → traced to `watchdog.php:179`, user-home-scoped, answer: no; (2) "do #121/#437 already cover the CPU cap?" → read both, answer: no. Both changed the output.

### Decomposition Targets

- **[REFACTOR]** `scripts/lib/user/watchdog.php` + `scripts/lib/user/userConfigStore.php` — converge the two enable/disable patterns (home-dir dotfile markers vs. operator-side JSON keys) onto the **operator-side JSON store**. This is the one genuine architectural cleanup #37 surfaces, and it is also the *abuse-relevant* choice: an operator-controlled flag in `/etc/seedbox/config/users/<user>.json` cannot be flipped by the user, unlike a `/home/<user>/.<marker>` dotfile. Track under #470's pattern decision.
- **[REFACTOR]** `scripts/cron/processWatchdog.sh:67` — the hardcoded `ffmpeg`/`HandBrakeCLI` name list is the existing instance of the rejected pattern. If miner mitigation ever needs a name component (defence-in-depth only, never the primary control), extend *this* one place rather than birthing a config-driven engine.

— Väinämöinen
